### ISSUE TITLE
Overload inliner regression in 0.50.0rc1

### ISSUE THREAD

**Issue by ehsantn** [numba/numba#5824]
## Reporting a bug

- [ ] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Overload inliner in 0.50.0rc1 reuses the implementation function's IR if the original function is called multiple times: https://github.com/numba/numba/blob/015dbf83df8b11bb25d0aa6b96cae809939dfaf6/numba/core/typed_passes.py#L681
Since Numba IR is mutable, reuse of IR nodes results in errors (I don't have a small reproducer yet). Therefore, a new IR should be generated each time (since IR cannot fully be copied yet).

**Comment by stuartarchibald** [comment]
Thanks for the report. Likely fixed by: https://github.com/numba/numba/pull/5812 ?

**Comment by ehsantn** [comment]
@stuartarchibald I don't believe #5812 actually solves the problem since the IR is not fully copied. Patching up a few places to avoid mutation is not enough, since the assumption that the IR is mutable is present everywhere (including user passes, not just Numba passes). Regenerating the IR each time is the safest option I believe.

**Comment by stuartarchibald** [comment]
The patch in #5812 solves it for the cases known but perhaps not for everything, suggests more tests are needed. I'm reluctant to re-run the entire untyped compilation pipeline every time a function is to be inlined as this is costly, I will look at copying the IR at the statement level just prior to inlining which ought to have the same effect. If you have a use case that triggers the issue then please could you share it? Thanks.

**Comment by ehsantn** [comment]
ParallelAccelerator uses `deepcopy` of IR already (e.g. stencils), but I'm not sure if it works for everything. Unfortunately, I can't create a small reproducer easily since it needs non-trivial optimization passes to trigger the issue.

**Comment by stuartarchibald** [comment]
This triggers it:
```python
from numba import njit, types
import numpy as np
from numba.core.extending import overload
from numba.core.typed_passes import InlineOverloads
from numba.core.compiler import CompilerBase, DefaultPassBuilder

class CustomCompiler(CompilerBase):

    def define_pipelines(self):
        pm = DefaultPassBuilder.define_nopython_pipeline(self.state)
        pm.add_pass_after(InlineOverloads, InlineOverloads)
        pm.finalize()
        return [pm]

def bar(x):
    ...

@overload(bar, inline='always')
def ol_bar(x):
    if isinstance(x, types.Integer):
        def impl(x):
            return x + 1.3
        return impl

@njit(pipeline_class=CustomCompiler)
def foo(z):
    return bar(z), bar(z)


foo(10)
```

**Comment by stuartarchibald** [comment]
I've patched https://github.com/numba/numba/pull/5812 as d5edcdf. Contains the deep copy from stencil source. Hopefully this will fix it, please could you test it against the bodo code base, thanks.

**Comment by ehsantn** [comment]
Thanks @stuartarchibald . This seems to work for my cases as expected.

**Comment by stuartarchibald** [comment]
Thanks for confirming @ehsantn. I really didn't want to lose the caching as recompilation of the same function each time it's inlined is an unnecessary expense!

**Comment by stuartarchibald** [comment]
Closing, fixed in #5812
